File: mcp-client-bridge/src/bridge-config.ts
/**
 * Configuration for the MCP Client Bridge.
 * Loaded from CLI args and environment variables.
 */
export interface BridgeConfig {
  orchestratorUrls: string[];
  orchestratorUrl: string;
  reconnectEnabled: boolean;
  maxReconnectDelayMs: number;
  baseReconnectDelayMs: number;
  requestTimeoutMs: number;
  connectionTimeoutMs: number;
  maxRetryBeforeRotate: number;
  enableLocalStreamWrite: boolean;
  enableLocalServers: boolean;
  configPath: string | undefined;
  healthIntervalMs: number;
  routingRefreshMs: number;
  pingIntervalMs: number;
  pingTimeoutMs: number;
  token: string | null;
}

export const BridgeConfig = {
  /**
   * Load configuration from CLI args and environment.
   * Priority: CLI args > env vars > defaults.
   */
  load(args: string[]): BridgeConfig {
    const token = parseToken(args);
    const urls = parseUrls(args);
    return {
      orchestratorUrls: urls,
      orchestratorUrl: urls[0],
      reconnectEnabled: !args.includes('--no-reconnect'),
      maxReconnectDelayMs: 15_000,
      baseReconnectDelayMs: 1_000,
      requestTimeoutMs: parseTimeout(args),
      connectionTimeoutMs: 5_000,
      maxRetryBeforeRotate: 3,
      enableLocalStreamWrite: !args.includes('--no-local-write'),
      enableLocalServers: !args.includes('--no-local-servers'),
      configPath: parseConfigPath(args),
      healthIntervalMs: parseHealthInterval(args),
      routingRefreshMs: parseRoutingRefresh(args),
      pingIntervalMs: parsePingInterval(args),
      pingTimeoutMs: parsePingTimeout(args),
      token,
    };
  },
};

/** Parse --token CLI arg or MCP_BRIDGE_TOKEN env. Validates JWT format. */
function parseToken(args: string[]): string | null {
  const idx = args.indexOf('--token');
  const cliToken = (idx >= 0 && idx + 1 < args.length) ? args[idx + 1] : null;
  const token = cliToken ?? process.env.MCP_BRIDGE_TOKEN ?? null;
  if (token) validateTokenFormat(token);
  return token;
}

/** Validate token is 3-part base64url JWT (header.payload.signature). */
function validateTokenFormat(token: string): void {
  const parts = token.split('.');
  if (parts.length !== 3) {
    console.error('[mcp-bridge] Invalid token format. Expected JWT (header.payload.signature)');
    process.exit(1);
  }
  const base64urlRegex = /^[A-Za-z0-9_-]+$/;
  for (const part of parts) {
    if (!base64urlRegex.test(part)) {
      console.error('[mcp-bridge] Invalid token format. JWT parts must be base64url encoded');
      process.exit(1);
    }
  }
}

function parseUrls(args: string[]): string[] {
  const idx = args.indexOf('--url');
  let raw: string;
  if (idx >= 0 && idx + 1 < args.length) {
    raw = args[idx + 1];
  } else {
    raw = process.env.ORCHESTRATOR_URLS
      ?? process.env.ORCHESTRATOR_URL
      ?? 'http://localhost:8080';
  }

  const urls = raw
    .split(',')
    .map(u => u.trim())
    .filter(u => u.length > 0)
    .filter(u => u.startsWith('http://') || u.startsWith('https://'));

  if (urls.length === 0) {
    console.error('[mcp-bridge] No valid URLs configured');
    process.exit(1);
  }
  if (urls.length > 10) {
    console.error('[mcp-bridge] URL list truncated to 10 entries');
    return urls.slice(0, 10);
  }
  return urls;
}

function parseTimeout(args: string[]): number {
  const idx = args.indexOf('--timeout');
  if (idx >= 0 && idx + 1 < args.length) {
    const val = parseInt(args[idx + 1], 10);
    if (!isNaN(val)) return val;
  }
  const envVal = process.env.BRIDGE_TIMEOUT;
  if (envVal) {
    const val = parseInt(envVal, 10);
    if (!isNaN(val)) return val;
  }
  return 30_000;
}

function parsePingInterval(args: string[]): number {
  const idx = args.indexOf('--ping-interval');
  if (idx >= 0 && idx + 1 < args.length) {
    const val = parseInt(args[idx + 1], 10);
    if (!isNaN(val) && (val === 0 || val >= 5000)) return val;
  }
  const envVal = process.env.BRIDGE_PING_INTERVAL;
  if (envVal) {
    const val = parseInt(envVal, 10);
    if (!isNaN(val) && (val === 0 || val >= 5000)) return val;
  }
  return 30_000;
}

function parsePingTimeout(args: string[]): number {
  const idx = args.indexOf('--ping-timeout');
  if (idx >= 0 && idx + 1 < args.length) {
    const val = parseInt(args[idx + 1], 10);
    if (!isNaN(val) && val >= 1000) return val;
  }
  const envVal = process.env.BRIDGE_PING_TIMEOUT;
  if (envVal) {
    const val = parseInt(envVal, 10);
    if (!isNaN(val) && val >= 1000) return val;
  }
  return 5_000;
}

function parseConfigPath(args: string[]): string | undefined {
  const idx = args.indexOf('--config-path');
  if (idx >= 0 && idx + 1 < args.length) return args[idx + 1];
  return process.env.MCP_CONFIG_PATH ?? undefined;
}

function parseHealthInterval(args: string[]): number {
  const idx = args.indexOf('--health-interval');
  if (idx >= 0 && idx + 1 < args.length) {
    const val = parseInt(args[idx + 1], 10);
    if (!isNaN(val) && val >= 0) return val;
  }
  const envVal = process.env.MCP_HEALTH_INTERVAL;
  if (envVal) {
    const val = parseInt(envVal, 10);
    if (!isNaN(val) && val >= 0) return val;
  }
  return 30_000;
}

function parseRoutingRefresh(args: string[]): number {
  const idx = args.indexOf('--routing-refresh');
  if (idx >= 0 && idx + 1 < args.length) {
    const val = parseInt(args[idx + 1], 10);
    if (!isNaN(val) && val >= 0) return val;
  }
  const envVal = process.env.MCP_ROUTING_REFRESH;
  if (envVal) {
    const val = parseInt(envVal, 10);
    if (!isNaN(val) && val >= 0) return val;
  }
  return 60_000;
}

File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/BridgeConfig.kt
package com.orchestrator.mcp.bridge

import org.slf4j.LoggerFactory

/**
 * Configuration for the MCP Client Bridge.
 * Loaded from CLI args and environment variables.
 */
data class BridgeConfig(
    val orchestratorUrls: List<String>,
    val orchestratorUrl: String = orchestratorUrls.first(),
    val reconnectEnabled: Boolean = true,
    val maxReconnectDelayMs: Long = 15_000,
    val baseReconnectDelayMs: Long = 1_000,
    val requestTimeoutMs: Long = 30_000,
    val connectionTimeoutMs: Long = 5_000,
    val maxRetryBeforeRotate: Int = 3,
    val enableLocalStreamWrite: Boolean = true,
    val pingIntervalMs: Long = 30_000,
    val pingTimeoutMs: Long = 5_000,
    val token: String? = null
) {
    companion object {
        private val logger = LoggerFactory.getLogger(BridgeConfig::class.java)
        private const val MAX_URLS = 10

        /**
         * Load configuration from CLI args and environment.
         * Priority: CLI args > env vars > defaults.
         */
        fun load(args: Array<String>): BridgeConfig {
            val urls = parseUrls(args)
            val timeout = parseTimeout(args)
            val noReconnect = args.contains("--no-reconnect")
            val pingInterval = parsePingInterval(args)
            val pingTimeout = parsePingTimeout(args)
            val token = parseToken(args)

            if (urls.size > 1) {
                logger.info("Bridge config: ${urls.size} URLs (failover), auth=${if (token != null) "JWT" else "NONE"}")
            } else {
                logger.info("Bridge config: url=${urls.first()}, auth=${if (token != null) "JWT" else "NONE"}")
            }
            return BridgeConfig(
                orchestratorUrls = urls,
                reconnectEnabled = !noReconnect,
                requestTimeoutMs = timeout,
                enableLocalStreamWrite = !args.contains("--no-local-write"),
                pingIntervalMs = pingInterval,
                pingTimeoutMs = pingTimeout,
                token = token
            )
        }

        private fun parseToken(args: Array<String>): String? {
            val idx = args.indexOf("--token")
            if (idx >= 0 && idx + 1 < args.size) return args[idx + 1]
            return System.getenv("MCP_BRIDGE_TOKEN")
        }

        private fun parseUrls(args: Array<String>): List<String> {
            val idx = args.indexOf("--url")
            val raw = when {
                idx >= 0 && idx + 1 < args.size -> args[idx + 1]
                System.getenv("ORCHESTRATOR_URLS") != null -> System.getenv("ORCHESTRATOR_URLS")
                System.getenv("ORCHESTRATOR_URL") != null -> System.getenv("ORCHESTRATOR_URL")
                else -> "http://localhost:8080"
            }
            val urls = raw.split(",")
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .filter { it.startsWith("http://") || it.startsWith("https://") }

            if (urls.isEmpty()) {
                logger.error("No valid URLs configured")
                System.exit(1)
            }
            if (urls.size > MAX_URLS) {
                logger.warn("URL list truncated to $MAX_URLS entries")
                return urls.take(MAX_URLS)
            }
            return urls
        }

        private fun parseTimeout(args: Array<String>): Long {
            val idx = args.indexOf("--timeout")
            if (idx >= 0 && idx + 1 < args.size) {
                return args[idx + 1].toLongOrNull() ?: 30_000
            }
            return System.getenv("BRIDGE_TIMEOUT")?.toLongOrNull()
                ?: 30_000
        }

        private fun parsePingInterval(args: Array<String>): Long {
            val idx = args.indexOf("--ping-interval")
            if (idx >= 0 && idx + 1 < args.size) {
                val v = args[idx + 1].toLongOrNull() ?: return 30_000
                if (v == 0L || v >= 5000) return v
            }
            val env = System.getenv("BRIDGE_PING_INTERVAL")?.toLongOrNull()
            if (env != null && (env == 0L || env >= 5000)) return env
            return 30_000
        }

        private fun parsePingTimeout(args: Array<String>): Long {
            val idx = args.indexOf("--ping-timeout")
            if (idx >= 0 && idx + 1 < args.size) {
                val v = args[idx + 1].toLongOrNull() ?: return 5_000
                if (v >= 1000) return v
            }
            val env = System.getenv("BRIDGE_PING_TIMEOUT")?.toLongOrNull()
            if (env != null && env >= 1000) return env
            return 5_000
        }
    }
}
